### PR TITLE
#162949216 User can create and list profiles

### PR DESCRIPTION
**Pull request by sylviawanjiku**
#### What does this PR do?

Create a User Profile on registration

#### Description of Task to be completed?

* Write tests to for the user profile 
* Create a user profile that displays: *username, bio, image/avatar*
* User should be able to update profile
* User can view other profiles

#### How should this be manually tested?

1. Clone the repo: `git clone https://github.com/andela/ah-the-unsullied.git`
2. Cd into the cloned repo 
3. Git checkout to branch ft-user-profile-162949216
4. Set up a virtual environment and install requirements:
      ```
            python3 -m venv venv
            source venv/bin/activate
            pip3 install -r requirements.txt

     ```
5. Set up a .env file in the same level as the venv file and add the following:
      ```
            source venv/bin/activate
            export SECRET_KEY='yoursecretkey'
            export DATABASE_URL='postgres://yourdatabasename'
            export DJANGO_SETTINGS_MODULE='authors.settings'
            export CLOUDINARY_NAME=' '
            export CLOUDINARY_KEY=' '
            export CLOUDINARY_SECRET=' '

      ```
6. Then run: 
     ``` 
            source .env 
            ./manage.py migrate
            ./manage.py runserver
      ```

7. Fire up postman and test the following endpoints:

| EndPoint                                        | Functionality                           |
| -------------------------------- |:----------------------------:|
| POST   api/users/                          | Create user profile                 |
| GET     api/profiles/ username     | Get single user profile            |
| PUT     api/profiles/ username     | Edit single user profile           |
| GET     api/profiles/                       | List all user profiles                | 

                                                    
use the following payload format to create a profile on registration:
```
{
  "user": {
    "email": "jake@jake.jake",
    "username": "jake",
    "password":"password"
  }
}
```

#### Any background context you want to provide?
You need cloudinary keys in a .env file to update the image field.

#### What are the relevant pivotal tracker stories?
[162949216](https://www.pivotaltracker.com/story/show/162949216)
[162949225](https://www.pivotaltracker.com/story/show/162949225)

#### Screenshots (if appropriate)
Get all profiles:

get_all

Get a single Profile:
detail

Update a Profile:
updated

### CODE REVIEW

File: authors/apps/authentication/views.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 from rest_framework.generics import CreateAPIView, RetrieveUpdateAPIView
 from rest_framework.permissions import AllowAny, IsAuthenticated
 from rest_framework.response import Response
+from rest_framework.views import APIView
```
Comment (nesh-dev) [inline authors/apps/authentication/views.py]: This import is not used in the file.

File: authors/apps/profiles/views.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+from rest_framework.generics import ListAPIView, RetrieveUpdateAPIView
+from rest_framework.renderers import JSONRenderer
```
Comment (nesh-dev) [inline authors/apps/profiles/views.py]: Import not used in file.

File: authors/apps/profiles/models.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+def create_profile_post_receiver(sender, instance, *args, **kwargs):
+    if kwargs['created']:
+        instance.user_profile = UserProfile.objects.create(user=instance)
+
+post_save.connect(create_profile_post_receiver, sender=User)
+
+def create_slug_pre_save_receiver(sender,instance, *args, **kwargs):
+    if not instance.slug:
+        instance.slug = unique_slug_generator(instance)
+
```
Comment (nesh-dev) [inline authors/apps/profiles/models.py]: PEP 8 two lines after the end of the class.

**Review by nesh-dev — CHANGES_REQUESTED** [review]

**Review by nesh-dev — DISMISSED** [review]

**Review by MbugwaSami — CHANGES_REQUESTED** [review]
Please remove the blank line at the end of models.py and also add another blank line after line 36 on the same file.

**Review by nesh-dev — DISMISSED** [review]

**Review by Allan690 — DISMISSED** [review]

**Review by nesh-dev — DISMISSED** [review]

**Review by kwanj-k — DISMISSED** [review]

**Review by MbugwaSami — DISMISSED** [review]

**Review by indungu — APPROVED** [review]
Nice work @sylviawanjiku and @kwanj-k

**Review by kwanj-k — APPROVED** [review]

**Pull request merged**